core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.kt

```diff
@@ -0,0 +1,677 @@
+    ) : this(makeAttrs(attrs), positions, speeds, timeDeltas)
+
+    /** Return the given metadata attribute */
+    @Suppress("UNCHECKED_CAST")
+    fun <T : SelfTypeHolder?> getAttr(attrType: Class<T>): T? {
+        return attrs[attrType] as T?
+    }
+
+    /** Returns whether this envelope part has a given attribute */
+    fun hasAttr(attrType: Class<out SelfTypeHolder>): Boolean {
+        return attrs.containsKey(attrType)
+    }
+
+    /**
+     * Returns whether the envelope has the given attribute value. Usually, we can't deduce the
+     * attribute type from the attribute value, but we can for enums.
+     */
+    fun <T : SelfTypeHolder?> hasAttr(attr: T): Boolean {
+        return attrs[attr!!.selfType] === attr
+    }
+
+    /** Returns a view of the envelope part attributes */
+    fun getAttrs(): Map<Class<out SelfTypeHolder>, SelfTypeHolder> {
+        return Collections.unmodifiableMap(attrs)
+    }
+
+    // endregion
+    // region SANITY_CHECKS
+    /**
+     * Runs every assertion on the envelope part values. <br></br> To be called in the constructor
+     * and after the values have been edited (which should be avoided when possible)
+     */
+    private fun runSanityChecks() {
+        assert(hasAttr(EnvelopeProfile::class.java)) { "missing EnvelopeProfile attribute" }
+        assert(positions.isNotEmpty()) { "attempted to create an empty EnvelopePart" }
+        assert(positions.size == speeds.size) {
+            "there must be the same number of point and speeds"
+        }
+        assert(timeDeltas.size == positions.size - 1) {
+            "there must be as many timeDeltas as gaps between points"
+        }
+        assert(checkNaNFree(positions)) { "NaNs in positions" }
+        assert(checkNaNFree(speeds)) { "NaNs in speeds" }
+        assert(checkNaNFree(timeDeltas)) { "NaNs in timeDeltas" }
+        assert(checkPositive(speeds)) { "negative speeds" }
+        assert(checkPositive(timeDeltas)) { "negative timeDeltas" }
+        assert(checkStrictlyMonotonicIncreasing(positions)) {
+            "positions aren't strictly increasing"
+        }
+    }
+
+    private fun checkPosition(stepIndex: Int, position: Double): Boolean {
+        return position >= getBeginPos(stepIndex) && position <= getEndPos(stepIndex)
+    }
+
+    // endregion
+    // region GETTERS
+    /** The number of points in the envelope part */
+    fun pointCount(): Int {
+        return positions.size
+    }
+
+    /** The number of steps in the envelope part */
+    fun stepCount(): Int {
+        return positions.size - 1
+    }
+
+    val maxSpeed: Double
+        /** Returns the maximum speed of the envelope part */
+        get() {
+            if (!java.lang.Double.isNaN(maxSpeedCache)) return maxSpeedCache
+
+            var maxSpeed = Double.NEGATIVE_INFINITY
+            for (speed in speeds) if (speed > maxSpeed) maxSpeed = speed
+            this.maxSpeedCache = maxSpeed
+            return maxSpeed
+        }
```

Comment:
```suggestion
    /** Returns the maximum speed of the envelope part */
    fun getMaxSpeed(): Double {
        if (!java.lang.Double.isNaN(maxSpeedCache)) return maxSpeedCache

        var maxSpeed = Double.NEGATIVE_INFINITY
        for (speed in speeds) if (speed > maxSpeed) maxSpeed = speed
        this.maxSpeedCache = maxSpeed
        return maxSpeed
    }
```

Same for other getters that aren't one-liners